### PR TITLE
JDK-8269387: jpackage --add-launcher should have option to not create shortcuts for additional launchers

### PR DESCRIPTION
**Pull request by andyherrick**
JDK-8269387: jpackage --add-launcher should have option to not create shortcuts for additional launchers

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8269387](https://bugs.openjdk.java.net/browse/JDK-8269387): jpackage --add-launcher should have option to not create shortcuts for additional launchers

### Reviewers
 * [Alexey Semenyuk](https://openjdk.java.net/census#asemenyuk) (@alexeysemenyukoracle - **Reviewer**)
 * [Alexander Matveev](https://openjdk.java.net/census#almatvee) (@sashamatveev - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4730/head:pull/4730` \
`$ git checkout pull/4730`

Update a local copy of the PR: \
`$ git checkout pull/4730` \
`$ git pull https://git.openjdk.java.net/jdk pull/4730/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4730`

View PR using the GUI difftool: \
`$ git pr show -t 4730`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4730.diff">https://git.openjdk.java.net/jdk/pull/4730.diff</a>

</details>

### CODE REVIEW

**Comment by kevinrushforth** [comment]
This will need a CSR (so you or someone with a Reviewer role should indicate this with the `/csr` command).

Also, can you summarize the changes in this PR, including the added command line options?

File: src/jdk.jpackage/share/classes/jdk/jpackage/internal/AddLauncherArguments.java [hunk 1]
```diff
@@ -120,6 +122,15 @@ private void initLauncherMap() {
         Arguments.putUnlessNull(bundleParams, CLIOptions.ICON.getId(),
                 (value == null) ? null : Path.of(value));
 
+        Arguments.putUnlessNull(bundleParams, SHORTCUT_HINT.getID(),
```
Comment (alexeysemenyukoracle) [inline src/jdk.jpackage/share/classes/jdk/jpackage/internal/AddLauncherArguments.java:125]: I think it would be better to add platform-specific options only if jpackage runs on that platform:
```
if (Platform.isWindows())  {
    Arguments.putUnlessNull(bundleParams, SHORTCUT_HINT.getID(), getOptionValue(CLIOptions.WIN_SHORTCUT_HINT));
    Arguments.putUnlessNull(bundleParams, MENU_HINT.getID(), getOptionValue(CLIOptions.WIN_MENU_HINT));
}

if (Platform.isLinux())  {
    Arguments.putUnlessNull(bundleParams, SHORTCUT_HINT.getID(), getOptionValue(CLIOptions.LINUX_SHORTCUT_HINT));
}

```

File: src/jdk.jpackage/share/classes/jdk/jpackage/internal/AddLauncherArguments.java [hunk 2]
```diff
@@ -32,6 +32,8 @@
 import jdk.jpackage.internal.Arguments.CLIOptions;
 import static jdk.jpackage.internal.StandardBundlerParam.LAUNCHER_DATA;
 import static jdk.jpackage.internal.StandardBundlerParam.APP_NAME;
+import static jdk.jpackage.internal.StandardBundlerParam.MENU_HINT;
+import static jdk.jpackage.internal.StandardBundlerParam.SHORTCUT_HINT;
 
 /*
  * AddLauncherArguments
```
Comment (alexeysemenyukoracle) [inline src/jdk.jpackage/share/classes/jdk/jpackage/internal/AddLauncherArguments.java:39]: Class comment needs to be updated. Currently it says:
```
 * The add-launcher properties file may have any of:
 *
 * appVersion
 * module
 * main-jar
 * main-class
 * icon
 * arguments
 * java-options
 * win-console
 * linux-app-category
```

File: src/jdk.jpackage/share/classes/jdk/jpackage/internal/AppImageFile.java [hunk 3]
```diff
@@ -226,10 +247,16 @@ public static Document readXml(Path appImageDir) throws IOException {
...
 
-        launchers.add(APP_NAME.fetchFrom(params));
+        List<String> names = new ArrayList<>();
+        names.add(APP_NAME.fetchFrom(params));
         ADD_LAUNCHERS.fetchFrom(params).stream().map(APP_NAME::fetchFrom).forEach(
-                launchers::add);
+                names::add);
+        for (var name : names) {
+            launchers.add(new LauncherInfo(name, true, true));
+        }
```
Comment (alexeysemenyukoracle) [inline src/jdk.jpackage/share/classes/jdk/jpackage/internal/AppImageFile.java:259]: This block of code can be simplified down to:
`ADD_LAUNCHERS.fetchFrom(params).stream().map(APP_NAME::fetchFrom).map(name -> new LauncherInfo(name, true, true)).forEach(launchers::add);`

File: test/jdk/tools/jpackage/share/AddLShortcutTest.java [hunk 4]
```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright (c) 2018, 2020, Oracle and/or its affiliates. All rights reserved.
```
Comment (alexeysemenyukoracle) [inline test/jdk/tools/jpackage/share/AddLShortcutTest.java:2]: Copyright year should be 2021

File: test/jdk/tools/jpackage/share/AddLShortcutTest.java [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+ * @requires (jpackage.test.SQETest == null)
+ * @library ../helpers
+ * @build jdk.jpackage.test.*
+ * @modules jdk.jpackage/jdk.jpackage.internal
+ * @compile AddLShortcutTest.java
+ * @run main/othervm/timeout=540 -Xmx512m jdk.jpackage.test.Main
+ *  --jpt-run=AddLShortcutTest
+ */
+
+public class AddLShortcutTest {
```
Comment (alexeysemenyukoracle) [inline test/jdk/tools/jpackage/share/AddLShortcutTest.java:72]: How does the test tests shortcuts?

File: test/jdk/tools/jpackage/share/AddLShortcutTest.java [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+
+/**
+ * Test --add-launcher parameter with shortcuts (platform permitting).
+ * Output of the test should be AddLShortcutTest*.* installer.
+ * The output installer should provide the same functionality as the
+ * default installer (see description of the default installer in
+ * SimplePackageTest.java) plus install three extra application launchers.
+ */
+
+/*
```
Comment (alexeysemenyukoracle) [inline test/jdk/tools/jpackage/share/AddLShortcutTest.java:46]: Why do you need two jtreg @test-s if there is only one test method in the test class? They are 100% duplicates.

**Review by alexeysemenyukoracle — CHANGES_REQUESTED** [review]
Review

**Review by alexeysemenyukoracle — CHANGES_REQUESTED** [review]

File: test/jdk/tools/jpackage/helpers/jdk/jpackage/test/AdditionalLauncher.java [hunk 7]
```diff
@@ -81,12 +81,14 @@ public AdditionalLauncher addRawProperties(
         return this;
     }
 
-    public AdditionalLauncher setShortcuts(boolean menu, boolean desktop) {
+    public AdditionalLauncher setShortcuts(boolean menu, boolean shortcut) {
+        withMenuShortcut = menu;
+        withShortcut = shortcut;
         if (TKit.isLinux()) {
```
Comment (alexeysemenyukoracle) [inline test/jdk/tools/jpackage/helpers/jdk/jpackage/test/AdditionalLauncher.java:87]: I'd move add `addRawProperties()` calls to `initialize()` method. This will prevent multiple `addRawProperties()` calls for "linux-shortcut" and "win-shortcut" properties in case `setShortcuts()` method is called multiple times.

**Review by alexeysemenyukoracle — CHANGES_REQUESTED** [review]

**Review by alexeysemenyukoracle — APPROVED** [review]

**Review by sashamatveev — APPROVED** [review]
